include/util/json_renderer.hpp

```diff
@@ -64,131 +78,94 @@ struct Renderer
-        out.insert(std::end(out), std::begin(string_to_insert), std::end(string_to_insert));
-        out.push_back('\"');
-    }
+template <> void Renderer<std::vector<char>>::write(const char *str)
+{
+    out.insert(out.end(), str, str + strlen(str));
+}
+template <> void Renderer<std::vector<char>>::write(char ch) { out.push_back(ch); }
 
-    void operator()(const Number &number) const
-    {
-        const std::string number_string = cast::to_string_with_precision(number.value);
-        out.insert(out.end(), number_string.begin(), number_string.end());
-    }
+template <> void Renderer<std::ostream>::write(const std::string &str) { out << str; }
 
-    void operator()(const Object &object) const
-    {
-        out.push_back('{');
-        for (auto it = object.values.begin(), end = object.values.end(); it != end;)
-        {
-            out.push_back('\"');
-            out.insert(out.end(), it->first.begin(), it->first.end());
-            out.push_back('\"');
-            out.push_back(':');
+template <> void Renderer<std::ostream>::write(const char *str) { out << str; }
 
-            mapbox::util::apply_visitor(ArrayRenderer(out), it->second);
-            if (++it != end)
-            {
-                out.push_back(',');
-            }
-        }
-        out.push_back('}');
-    }
+template <> void Renderer<std::ostream>::write(char ch) { out << ch; }
 
-    void operator()(const Array &array) const
-    {
-        out.push_back('[');
-        for (auto it = array.values.cbegin(), end = array.values.cend(); it != end;)
-        {
-            mapbox::util::apply_visitor(ArrayRenderer(out), *it);
-            if (++it != end)
-            {
-                out.push_back(',');
-            }
-        }
-        out.push_back(']');
-    }
+template <> void Renderer<std::string>::write(const std::string &str) { out += str; }
 
-    void operator()(const True &) const
-    {
-        const std::string temp("true");
-        out.insert(out.end(), temp.begin(), temp.end());
-    }
+template <> void Renderer<std::string>::write(const char *str) { out += str; }
 
-    void operator()(const False &) const
-    {
-        const std::string temp("false");
-        out.insert(out.end(), temp.begin(), temp.end());
-    }
-
-    void operator()(const Null &) const
-    {
-        const std::string temp("null");
-        out.insert(out.end(), temp.begin(), temp.end());
-    }
-
-  private:
-    std::vector<char> &out;
-};
+template <> void Renderer<std::string>::write(char ch) { out += ch; }
 
 inline void render(std::ostream &out, const Object &object)
 {
-    Value value = object;
-    mapbox::util::apply_visitor(Renderer(out), value);
+    Renderer renderer(out);
+    renderer(object);
+}
+
+inline void render(std::string &out, const Object &object)
+{
+    Renderer renderer(out);
+    renderer(object);
 }
 
 inline void render(std::vector<char> &out, const Object &object)
 {
-    Value value = object;
```

Comment:
# 3.
And finally I noticed that we are doing unneeded copy of the whole json here and it gave the most significant improvement:
<img width="853" alt="get rid of copy" src="https://user-images.githubusercontent.com/266271/193302350-11dbff42-0516-4c7a-a28e-d2ea615d581f.png">